Assets/MixedRealityToolkit/Providers/Hands/BaseHandVisualizer.cs

```diff
@@ -18,14 +18,31 @@ public class BaseHandVisualizer : MonoBehaviour, IMixedRealityHandVisualizer, IM
         protected readonly Dictionary<TrackedHandJoint, Transform> joints = new Dictionary<TrackedHandJoint, Transform>();
         protected MeshFilter handMeshFilter;
 
+        private IMixedRealityInputSystem inputSystem = null;
+
+        /// <summary>
+        /// The active instance of the input system.
+        /// </summary>
+        protected IMixedRealityInputSystem InputSystem
+        {
+            get
+            {
+                if (inputSystem == null)
+                {
+                    MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem);
+                }
+                return inputSystem;
+            }
+        }
+
```

Comment:
This overall change looks fine, it's is weird to me though that we go from one singleton (i.e. MixedRealityToolkit) to another (MixedRealityServiceRegistry) - in this case now we put the onus of caching and reaching out to all of the consumers (i.e. this exact code is copy pasted in a lot of the things here)

If every single caller is handling caching it almost feels like MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem> should cache as well (i.e. why not share some of this code at a lower level)

Reply:
> we go from one singleton (i.e. MixedRealityToolkit) to another (MixedRealityServiceRegistry)

MixedRealityServiceRegistry is a static c# class that has no scene presence (it is not a Unity object).

> If every single caller is handling caching it almost feels like MixedRealityServiceRegistry.TryGetService should cache as well (i.e. why not share some of this code at a lower level)

Interesting thought. This is actually very similar to the original pattern where many scripts would attempt to get the service if a previous instance had not yet been found (ex: was null)

Reply:
As an aside, caching is an optimization and is not required. Any script can request a system at the time it is needed.